### ISSUE TITLE
[BUG] Memory leak using zero stage 3, CUDA out of memory after training several minutes

### ISSUE THREAD

**Issue by taoisu** [microsoft/DeepSpeed#2057]
**Describe the bug**

I'm using pytorch_lightning with deepspeed to train a ddpm model, the training started with ~6G GPU memory used, but it keeps increasing as the training continues, after about ~100 steps the memory usage increased to more than 11G, causing CUDA out of memory issue on my 2080ti card.

**To Reproduce**
1. Git clone CLIP repo & install: `git clone git@github.com:taoisu/CLIP.git; cd CLIP; pip install -e .;`
2. Git clone latent diffusion repo & install: `git clone git@github.com:taoisu/latent-diffusion.git; cd latent-diffusion; git checkout xiaguo/ds-debug; pip install -e .;`
3. Run the training with dummy data: `cd latent-diffusion; python main.py --base configs/just-diffusion/pin64_ds.yaml --gpus 0, -t;`
4. Check GPU memory usage

**Expected behavior**
Not expecting CUDA out of memory issue to happen

**ds_report output**
Please run `ds_report` to give us details about your setup.
```
--------------------------------------------------
DeepSpeed C++/CUDA extension op report
--------------------------------------------------
NOTE: Ops not installed will be just-in-time (JIT) compiled at
      runtime if needed. Op compatibility means that your system
      meet the required dependencies to JIT install the op.
--------------------------------------------------
JIT compiled ops requires ninja
ninja .................. [OKAY]
--------------------------------------------------
op name ................ installed .. compatible
--------------------------------------------------
cpu_adam ............... [NO] ....... [OKAY]
cpu_adagrad ............ [NO] ....... [OKAY]
fused_adam ............. [NO] ....... [OKAY]
fused_lamb ............. [NO] ....... [OKAY]
 [WARNING]  please install triton==1.0.0 if you want to use sparse attention
sparse_attn ............ [NO] ....... [NO]
transformer ............ [NO] ....... [OKAY]
stochastic_transformer . [NO] ....... [OKAY]
 [WARNING]  async_io requires the dev libaio .so object and headers but these were not found.
 [WARNING]  async_io: please install the libaio-dev package with apt
 [WARNING]  If libaio is already installed (perhaps from source), try setting the CFLAGS and LDFLAGS environment variables to where it can be found.
async_io ............... [NO] ....... [NO]
utils .................. [NO] ....... [OKAY]
quantizer .............. [NO] ....... [OKAY]
transformer_inference .. [NO] ....... [OKAY]
--------------------------------------------------
DeepSpeed general environment info:
torch install path ............... ['/home/xiaguo/miniconda3/envs/nft-py38/lib/python3.8/site-packages/torch']
torch version .................... 1.11.0
torch cuda version ............... 11.3
torch hip version ................ None
nvcc version ..................... 11.6
deepspeed install path ........... ['/home/xiaguo/miniconda3/envs/nft-py38/lib/python3.8/site-packages/deepspeed']
deepspeed info ................... 0.6.5, unknown, unknown
deepspeed wheel compiled w. ...... torch 1.11, cuda 11.3
```

**Screenshots**

Training start:
<img width="562" alt="image" src="https://user-images.githubusercontent.com/30015018/175852971-352dfb13-584e-4e06-9594-46383c04303e.png">

<img width="1101" alt="image" src="https://user-images.githubusercontent.com/30015018/175853048-bf2266c0-6922-4f67-bc6b-626bd1924b0b.png">

After a while:

<img width="591" alt="image" src="https://user-images.githubusercontent.com/30015018/175854200-91c3f1c1-5dc0-4642-a28c-797b187492e0.png">

<img width="1101" alt="image" src="https://user-images.githubusercontent.com/30015018/175854171-ee06a4f2-29a9-4f18-a331-db182711af6d.png">

**System info (please complete the following information):**
 - OS: Ubuntu 18.04
 - GPU count and types 1 machine with 1x2080ti
 - Interconnects (if applicable) [e.g., two machines connected with 100 Gbps IB]
 - Python version 3.8.12
 - Any other relevant info about your setup

**Launcher context**
Are you launching your experiment with the `deepspeed` launcher, MPI, or something else?

Launching it directly locally with python interpreter

**Docker context**
Are you using a specific docker image that you can share? No

**Additional context**

Searching online people seem to be able to mitigate this issue by calling `torch.cuda.empty_cache()` in the training loop, I tried that in the code but not helping. In fact the code is in the repro branch mentioned above.

deepspeed stage is specified in the pin64_ds.yaml file.

**Comment by GuanhuaWang** [comment]
Hi @taoisu , one possible issue I do notice is your system cuda version is 11.6, but your deepspeed is compiled using cuda 11.3, which may cause the issue of memory leak.

I tried reproducing the issue above but failed. I installed library dependency as `fairscale`, `pytorch_lightning`, `omegaconf`, `einops`. There is no `xiaguo/ds-debug` branch in your `latent-diffusion` repo. I tried to run with `main` branch and it reports following errors.:

```
Traceback (most recent call last):
  File "main.py", line 865, in <module>
    if trainer.global_rank == 0:
NameError: name 'trainer' is not defined
```
